File: cpp/2024/day01/main.cpp
#include <algorithm>
#include <cstddef>
#include <cstdint>
#include <iostream>
#include <string>
#include <vector>

#include "../../utils/EggUtils.h"

int main() {
	// std::vector<std::string> input = egg::readInputFile("./inputTest.txt");
	std::vector<std::string> input = egg::readInputFile("./input.txt");

	std::vector<int> rightList;
	std::vector<int> leftList;
	rightList.reserve(input.size());
	leftList.reserve(input.size());

	for(std::string line : input) {
		std::vector<std::string> nums = egg::string_split(line, "   ");
		leftList.emplace_back(std::stoi(nums.at(0)));
		rightList.emplace_back(std::stoi(nums.at(1)));
	}

	std::uint64_t sum = 0;
	for(std::size_t i = 0; i < rightList.size(); i++) {
		auto times = std::count(rightList.begin(), rightList.end(), leftList.at(i));
		sum += leftList.at(i) * times;
	}

	std::cout << "sum: " << sum << "\n";

	return 0;
}

File: zig/2024/day01/main.zig
const std = @import("std");
const egg = @import("./EggUtils.zig");

pub fn main() !void {
    const input = try egg.readLines(std.heap.page_allocator, "./input.txt");
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var rightList = try allocator.alloc(i32, input.items.len);
    defer allocator.free(rightList);

    var leftList = try allocator.alloc(i32, input.items.len);
    defer allocator.free(leftList);

    for (input.items, 0..) |line, i| {
        const pos = std.mem.indexOf(u8, line, " ") orelse 0;

        rightList[i] = try std.fmt.parseInt(i32, line[0..pos], 10);
        leftList[i] = try std.fmt.parseInt(i32, line[pos + 1 .. line.len], 10);
        std.debug.print("{d}\n", .{rightList[i]});
    }

    var sum: i64 = 0;
    for (leftList) |l| {
        sum += l * count(rightList, l);
    }

    std.debug.print("sum: {d}\n", .{sum});
}

pub fn count(haystack: []i32, needle: i32) i32 {
    var sum: i32 = 0;
    for (haystack) |n| {
        sum += if (n == needle) 1 else 0;
    }
    return sum;
}
